### ISSUE TITLE
Adding an environment.yml

### ISSUE THREAD

**Issue by mistakevin** [bokeh/bokeh#2624]
Are we interested in adding one of these? It would make environment managing a lot easier, help new contributors set up quicker (just run `conda env create -f environment.yml`), and probably be utilized in the build process for testing, etc. It has some defaults from the `conda create --name bokeh python=2.7` that we can remove. Also, it will probably change for a python 3.4 environment.
```yaml
name: bokeh
channels:
- anaconda
- asmeurer
- bokeh
dependencies:
- alabaster=0.7.3=py27_0
- babel=1.3=py27_0
- beautiful-soup=4.3.2=py27_0
- bokeh=0.9.2=np19py27_0
- boto=2.38.0=py27_0
- brewer2mpl=1.4=py27_0
- certifi=14.05.14=py27_0
- colorama=0.3.3=py27_0
- coverage=3.7.1=py27_0
- docutils=0.12=py27_0
- flask=0.10.1=py27_1
- freetype=2.5.2=2
- funcsigs=0.4=py27_0
- gevent=1.0.1=py27_0
- gevent-websocket=0.9.3=py27_0
- ggplot=0.6.5=np19py27_0
- greenlet=0.4.7=py27_0
- husl=4.0.0=py27_0
- itsdangerous=0.24=py27_0
- jinja2=2.8=py27_0
- libpng=1.6.17=0
- libsodium=0.4.5=0
- markupsafe=0.23=py27_0
- matplotlib=1.4.3=np19py27_3
- mock=1.2.0=py27_0
- ncurses=5.9=1
- nose=1.3.7=py27_0
- numpy=1.9.2=py27_0
- openssl=1.0.1k=1
- pandas=0.16.2=np19py27_0
- patsy=0.4.0=np19py27_0
- pbr=1.3.0=py27_0
- pip=7.1.0=py27_0
- pygments=2.0.2=py27_0
- pyparsing=2.0.3=py27_0
- pyqt=4.11.3=py27_0
- python=2.7.10=0
- python-dateutil=2.4.2=py27_0
- pytz=2015.4=py27_0
- pyyaml=3.11=py27_1
- pyzmq=14.7.0=py27_0
- qt=4.8.6=3
- readline=6.2=2
- requests=2.7.0=py27_0
- scipy=0.16.0=np19py27_1
- seaborn=0.6.0=np19py27_0
- setuptools=18.0.1=py27_0
- sip=4.16.5=py27_0
- six=1.9.0=py27_0
- snowballstemmer=1.2.0=py27_0
- sphinx=1.3.1=py27_0
- sphinx-bootstrap-theme=0.4.2=py27_0
- sphinx_rtd_theme=0.1.7=py27_0
- sphinxcontrib-httpdomain=1.3.0=py27_0
- sphinxcontrib-napoleon=0.2.8=py27_0
- sqlite=3.8.4.1=1
- ssl_match_hostname=3.4.0.2=py27_0
- statsmodels=0.6.1=np19py27_0
- tk=8.5.18=0
- tornado=4.2.1=py27_0
- ujson=1.33=py27_0
- websocket-client=0.12.0=py27_1
- werkzeug=0.10.4=py27_0
- yaml=0.1.6=0
- zeromq=4.0.5=0
- zlib=1.2.8=0
- pip:
  - backports.ssl-match-hostname==3.4.0.2
  - beautifulsoup4==4.3.2
  - graphviz==0.4.5
  - pycallgraph==1.0.1
  - sphinx-rtd-theme==0.1.7
```

**Comment by mattpap** [comment]
I would be definitively interested, but this duplication bothers me. `conda env` should be able to take advantage of conda's metadata. Besides this, assuming we don't improve `conda env`, this list of packages has very specific version numbers. This diverges from `meta.yaml`. Is this intentional? If so, why?

**Comment by damianavila** [comment]
I am also interested but worried about the duplication too... We should work out a solution in a unique place and able to compose the `meta.yaml` files, the `env.yaml` file and the `setup.py` file as well...

**Comment by canavandl** [comment]
I'm unclear where meta.yaml comes from. Is that generated from setup.py or what?

**Comment by canavandl** [comment]
I'm +1 on having some sort of environment.yaml or conda channel to make the dev set-up smoother tho.

**Comment by mistakevin** [comment]
I also don't know what I don't know what _meta.yml_ or _env.yml_ are.

The specific versioning of the dependencies were generated by `conda env export > environment.yml`, which just exports the exact versions you have in your current environment. I can not find any documentation on the specifics of _environment.yml_, but hopefully there is variable dependency versioning like in [package.json](https://docs.npmjs.com/files/package.json#dependencies).

Just curious, how does testing on TravisCI work? Does it boot up a clean environment and download all dependencies before testing, as a new contributor would, or does the machine have pre-installed dependencies and maintain some sort of continuity between builds.

**Comment by damianavila** [comment]
>I also don't know what I don't know what meta.yml or env.yml are.

`meta` in inside the conda recipe in the repo, `env` is the output of conda env as you said below.

I am proposing to only have `meta` and being able to read that from `setup.py` and also maybe something to translate programmatically the `meta` file into a conda env-readable file to use environment set up support

>I can not find any documentation on the specifics of environment.yml, but hopefully there is variable dependency versioning like in package.json.

You will probably find documentation about the versioning scheme in the conda docs...

>Just curious, how does testing on TravisCI work? Does it boot up a clean environment and download all dependencies before testing, as a new contributor would, or does the machine have pre-installed dependencies and maintain some sort of continuity between builds.

Clean environment every time and installation of all the dependencies: https://github.com/bokeh/bokeh/blob/master/scripts/travis_install

**Comment by mistakevin** [comment]
Would a new contributor be able to leverage the `meta` file into an environment setup file?

**Comment by mattpap** [comment]
First step in #2626. I started working on conda-env to support `meta.yaml`.

**Comment by mistakevin** [comment]
Cool. And here is an example I was able to track down using [non-fixed versions of packages in an _environment.yml_](https://github.com/conda/conda.io/blob/master/environment.yml) if anyone was interested.

**Comment by tonyfast** [comment]
Closed by #2880
